// motion-v/dist/es/index.mjs
import { createContext } from "./utils/createContext.mjs";
import { resolveMotionProps } from "./utils/resolve-motion-props.mjs";
import { useInView } from "./utils/use-in-view.mjs";
import { useAnimationFrame } from "./utils/use-animation-frame.mjs";
import { getContextWindow } from "./utils/get-context-window.mjs";
import { useDomRef } from "./utils/use-dom-ref.mjs";
import { usePageInView } from "./utils/use-page-in-view.mjs";
import { injectLayoutGroup, injectMotion, layoutGroupInjectionKey, motionInjectionKey, provideLayoutGroup, provideMotion } from "./components/context.mjs";
import { provideMotionConfig, useMotionConfig } from "./components/motion-config/context.mjs";
import MotionConfig_default from "./components/motion-config/MotionConfig.mjs";
import { motionGlobalConfig } from "./config.mjs";
import { mountedStates } from "./state/motion-state.mjs";
import { domMax } from "./features/dom-max.mjs";
import { Motion, motion } from "./components/motion/index.mjs";
import AnimatePresence_default from "./components/animate-presence/AnimatePresence.mjs";
import { useCombineMotionValues } from "./value/use-combine-values.mjs";
import { useComputed } from "./value/use-computed.mjs";
import { useTransform } from "./value/use-transform.mjs";
import { useTime } from "./value/use-time.mjs";
import { useMotionTemplate } from "./value/use-motion-template.mjs";
import { useMotionValueEvent } from "./value/use-motion-value-event.mjs";
import { useSpring } from "./value/use-spring.mjs";
import { useScroll } from "./value/use-scroll.mjs";
import { useVelocity } from "./value/use-velocity.mjs";
import { Reorder, ReorderGroup, ReorderItem } from "./components/reorder/index.mjs";
import RowValue_default from "./components/RowValue.mjs";
import { LazyMotion } from "./components/lazy-motion/index.mjs";
import { M, m } from "./components/motion/m.mjs";
import { useLayoutGroup } from "./components/use-layout-group.mjs";
import LayoutGroup_default from "./components/LayoutGroup.mjs";
import { useAnimate } from "./animation/hooks/use-animate.mjs";
import { useAnimationControls } from "./animation/hooks/use-animation-controls.mjs";
import { useReducedMotion } from "./animation/hooks/use-reduced-motion.mjs";
import { useDragControls } from "./features/gestures/drag/use-drag-controls.mjs";
import { domAnimation } from "./features/dom-animation.mjs";
import { MotionPlugin, createMotionDirective, createPresetDirective, vMotion } from "./directive/index.mjs";
import { addScaleCorrector, delay as delayInMs, motionValue as useMotionValue } from "motion-dom";
export * from "framer-motion/dom";
export { AnimatePresence_default as AnimatePresence, LayoutGroup_default as LayoutGroup, LazyMotion, M, Motion, MotionConfig_default as MotionConfig, MotionPlugin, Reorder, ReorderGroup, ReorderItem, RowValue_default as RowValue, addScaleCorrector, createContext, createMotionDirective, createPresetDirective, delayInMs, domAnimation, domMax, getContextWindow, injectLayoutGroup, injectMotion, layoutGroupInjectionKey, m, motion, motionGlobalConfig, motionInjectionKey, mountedStates, provideLayoutGroup, provideMotion, provideMotionConfig, resolveMotionProps, useAnimate, useAnimationControls, useAnimationFrame, useCombineMotionValues, useComputed, useDomRef, useDragControls, useInView, useLayoutGroup, useMotionConfig, useMotionTemplate, useMotionValue, useMotionValueEvent, usePageInView, useReducedMotion, useScroll, useSpring, useTime, useTransform, useVelocity, vMotion };
